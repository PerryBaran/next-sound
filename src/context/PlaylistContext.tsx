'use client'

import { createContext, useContext, useState } from 'react'

interface Song {
  image?: string
  artistName: string
  albumName: string
  songName: string
  audio: string
}

interface Playlist extends Song {
  key: string
}

const PlaylistContext = createContext({
  playlist: [] as Playlist[],
  playlistIndex: 0,
  skipSong: (i: number | boolean) => {},
  handleAddToPlaylist: (
    songs: Song[],
    addNext: boolean,
    playNow: boolean
  ) => {},
  removeFromPlaylist: (i: number) => {},
  playing: false,
  handlePlaying: (isPlaying: boolean | undefined) => {}
})

const usePlaylistContext = () => useContext(PlaylistContext)

function PlaylistProvider({ children }: { children: React.ReactNode }) {
  const [playlist, setPlaylist] = useState<Playlist[]>([])
  const [playlistIndex, setPlaylistIndex] = useState(0)
  const [playing, setPlaying] = useState(false)

  const addToPlaylist = (songs: Song[], addNext = false) => {
    const data = songs.map((song) => {
      return { ...song, key: crypto.randomUUID() }
    })

    setPlaylist((prev) => {
      if (addNext) {
        const clone = [...prev]
        clone.splice(playlistIndex + 1, 0, ...data)
        return clone
      }
      return [...prev, ...data]
    })
  }

  const handlePlayNow = () => {
    if (playlist.length !== 0) {
      setPlaylistIndex((prev) => prev + 1)
    }
    setPlaying(true)
  }

  const handleAddToPlaylist = (
    songs: Song[],
    addNext: boolean,
    playNow: boolean
  ) => {
    addToPlaylist(songs, addNext)
    if (playNow) {
      handlePlayNow()
    }
  }

  const removeFromPlaylist = (i: number) => {
    setPlaylist((prev) => {
      const clone = [...prev]
      clone.splice(i, 1)
      return clone
    })

    setPlaylistIndex((prev) => {
      if (i > prev || prev === 0) {
        return prev
      }
      return prev - 1
    })
  }

  const skipSong = (value: number | boolean) => {
    const lastPlaylistIndex = playlist.length - 1
    if (typeof value === 'number') {
      if (value > lastPlaylistIndex) {
        setPlaylistIndex(lastPlaylistIndex)
      } else if (value < 0) {
        setPlaylistIndex(0)
      } else {
        setPlaylistIndex(value)
      }
    } else if (value) {
      if (playlistIndex === lastPlaylistIndex) {
        setPlaylistIndex(0)
      } else {
        setPlaylistIndex((prev) => prev + 1)
      }
    } else {
      if (playlistIndex > 0) {
        setPlaylistIndex((prev) => prev - 1)
      } else {
        setPlaylistIndex(lastPlaylistIndex)
      }
    }
  }

  const handlePlaying = (isPlaying: boolean | undefined) => {
    if (isPlaying === undefined) {
      setPlaying((prev) => !prev)
    } else {
      setPlaying(isPlaying)
    }
  }

  return (
    <PlaylistContext.Provider
      value={{
        playlist,
        playlistIndex,
        handleAddToPlaylist,
        removeFromPlaylist,
        skipSong,
        playing,
        handlePlaying
      }}
    >
      {children}
    </PlaylistContext.Provider>
  )
}

export { usePlaylistContext, PlaylistProvider }
