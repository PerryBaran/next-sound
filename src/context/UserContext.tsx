'use client'

import { createContext, useContext, useState, useEffect } from 'react'
import { getUserById } from '@/requests/users'
import Cookie from 'js-cookie'
import jwtDecode from 'jwt-decode'

interface User {
  name?: string
  id?: string
}

const emptyUser = {
  name: '',
  id: ''
}

const UserContext = createContext({
  user: emptyUser,
  handleLogin: (data: User | undefined) => {}
})

const useUserContext = () => useContext(UserContext)

function UserProvider({ children }: { children: React.ReactNode }) {
  const [user, setUser] = useState(emptyUser)

  const handleLogin = (data: User | undefined) => {
    if (data === undefined) {
      Cookie.remove('userToken')
      setUser(emptyUser)
    } else {
      setUser((prev) => {
        return { ...prev, ...data }
      })
    }
  }

  useEffect(() => {
    ;(async () => {
      const token = Cookie.get('userToken')
      if (!token) return
      const { id }: { id: string } = jwtDecode(token)
      const response = await getUserById(id)
      handleLogin({
        name: response.name,
        id
      })
    })()
  }, [])

  return (
    <UserContext.Provider value={{ user, handleLogin }}>
      {children}
    </UserContext.Provider>
  )
}

export { useUserContext, UserProvider }
