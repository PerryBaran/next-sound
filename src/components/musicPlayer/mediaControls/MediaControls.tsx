'use client'

import Image from 'next/image'
import style from './mediaControls.module.css'
import { play, pause, skip } from '@/media/icons'
import { usePlaylistContext } from '@/context/PlaylistContext'

interface Props {
  playing: boolean
  skipSong: (value: number | boolean) => void
}

export default function MediaControls(props: Props) {
  const { playing, skipSong } = props
  const { handlePlaying } = usePlaylistContext()

  return (
    <div className={style.container}>
      <button onClick={() => skipSong(false)}>
        <Image
          src={skip}
          alt="skip backwards"
          height={20}
          width={20}
          style={{ transform: 'rotate(180deg)' }}
        />
      </button>
      <button onClick={() => handlePlaying(undefined)}>
        {playing ? (
          <Image src={pause} alt="pause" height={20} width={20} />
        ) : (
          <Image src={play} alt="play" height={20} width={20} />
        )}
      </button>
      <button onClick={() => skipSong(true)}>
        <Image src={skip} alt="skip forwards" height={20} width={20} />
      </button>
    </div>
  )
}
