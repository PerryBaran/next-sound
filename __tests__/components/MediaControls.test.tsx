import { fireEvent, render, screen } from "@testing-library/react"
import { createContext, useContext } from "react"
import MediaControls from "../../src/components/musicPlayer/mediaControls/MediaControls"
import * as PlaylistContext from "../../src/context/PlaylistContext"

describe("MediaControls", () => {
  let props = {
    playing: true,
    skipSong: jest.fn()
  }
  let mockHandlePlaying = jest.fn()

  beforeEach(() => {
    props.skipSong = jest.fn()
    mockHandlePlaying = jest.fn()

    const mockContext = createContext({
      playlist: [],
      playlistIndex: 0,
      skipSong: jest.fn(),
      handleAddToPlaylist: jest.fn(),
      removeFromPlaylist: jest.fn(),
      playing: false,
      handlePlaying: mockHandlePlaying,
      handleShuffle: jest.fn()
    })

    jest
      .spyOn(PlaylistContext, "usePlaylistContext")
      .mockImplementation(() => useContext(mockContext))
  })

  test("renders correctly when playing is true", () => {
    render(<MediaControls {...props} />)

    expect(screen.getAllByRole("button")).toHaveLength(3)
    expect(screen.getByAltText("skip backwards")).toBeInTheDocument()
    expect(screen.getByAltText("pause")).toBeInTheDocument()
    expect(screen.getByAltText("skip forwards")).toBeInTheDocument()
  })

  test("renders correctly when playing is false", () => {
    render(<MediaControls {...props} playing={false} />)

    expect(screen.getAllByRole("button")).toHaveLength(3)
    expect(screen.getByAltText("skip backwards")).toBeInTheDocument()
    expect(screen.getByAltText("play")).toBeInTheDocument()
    expect(screen.getByAltText("skip forwards")).toBeInTheDocument()
  })

  test("skip backwards click", () => {
    render(<MediaControls {...props} />)

    expect(props.skipSong).not.toBeCalled()

    fireEvent.click(screen.getByAltText("skip backwards"))

    expect(props.skipSong).toBeCalledTimes(1)
    expect(props.skipSong).toBeCalledWith(false)
  })

  test("skip forwards click", () => {
    render(<MediaControls {...props} />)

    expect(props.skipSong).not.toBeCalled()

    fireEvent.click(screen.getByAltText("skip forwards"))

    expect(props.skipSong).toBeCalledTimes(1)
    expect(props.skipSong).toBeCalledWith(true)
  })

  test("play click", () => {
    render(<MediaControls {...props} />)

    expect(mockHandlePlaying).toBeCalledTimes(0)

    fireEvent.click(screen.getByAltText("pause"))

    expect(mockHandlePlaying).toBeCalledTimes(1)
  })

  test("pause click", () => {
    render(<MediaControls {...props} playing={false} />)

    expect(mockHandlePlaying).toBeCalledTimes(0)

    fireEvent.click(screen.getByAltText("play"))

    expect(mockHandlePlaying).toBeCalledTimes(1)
  })

  describe("snapshot", () => {
    test("playing is true", () => {
      const { asFragment } = render(<MediaControls {...props} />)

      expect(asFragment()).toMatchSnapshot()
    })

    test("playing is false", () => {
      const { asFragment } = render(
        <MediaControls {...props} playing={false} />
      )

      expect(asFragment()).toMatchSnapshot()
    })
  })
})
