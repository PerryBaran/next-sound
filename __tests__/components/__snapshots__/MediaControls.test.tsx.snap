// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaControls snapshot playing is false 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <button>
      <img
        alt="skip backwards"
        data-nimg="1"
        decoding="async"
        height="20"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
        style="color: transparent; transform: rotate(180deg);"
        width="20"
      />
    </button>
    <button>
      <img
        alt="play"
        data-nimg="1"
        decoding="async"
        height="20"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
        style="color: transparent;"
        width="20"
      />
    </button>
    <button>
      <img
        alt="skip forwards"
        data-nimg="1"
        decoding="async"
        height="20"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
        style="color: transparent;"
        width="20"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`MediaControls snapshot playing is true 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <button>
      <img
        alt="skip backwards"
        data-nimg="1"
        decoding="async"
        height="20"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
        style="color: transparent; transform: rotate(180deg);"
        width="20"
      />
    </button>
    <button>
      <img
        alt="pause"
        data-nimg="1"
        decoding="async"
        height="20"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
        style="color: transparent;"
        width="20"
      />
    </button>
    <button>
      <img
        alt="skip forwards"
        data-nimg="1"
        decoding="async"
        height="20"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
        style="color: transparent;"
        width="20"
      />
    </button>
  </div>
</DocumentFragment>
`;
